### DISCUSSION TITLE
Freeze the list of reserved words for stash

### DISCUSSION THREAD

**Discussion by polettix** [mojolicious/mojo#2081] [Ideas]
As of version 9.22, the docs for method `stash` have this:

> Some stash values have a special meaning and are reserved, the full list is currently `action`, `app`, \[...\]

*Currently* is scary: it opens the door to *any other* word to be included in the list in the future, potentially breaking backwards compatibility for applications.

As I understand from some comments in #1889, the list should be stable at this point, because Mojolicious is pretty mature. So adding a new reserved word would be a quite rare event, but still possible as implied by the docs.

If the `mojo.*` sub-namespace can be used for future extensions to the list, I think it's fair to freeze the list of reserved words leaving the note about `mojo.*`. Otherwise, if `mojo.*` is really *internal*, I would propose to allocate a similar prefix for these extensions and add it to the list of reserved words, *then* freeze that list.

All in all, my suggestion is to freeze the list of reserved words and get rid of the *currently* in a way or another.

**Comment by kraih** [comment]
The `mojo.*` namespace is completely internal and should not be used outside of core Mojolicious. Not even by plugins.

**Comment by kraih** [comment]
If we freeze the list, it would kinda mark mojo.js as the more actively maintained framework. Since it would still be able to add new rendering features freely, while Mojolicious could not anymore. I don't think that's a good idea.

**Comment by polettix** [comment]
Freezing the set of top-level reserved words in the stash that could clash with user keys/variables does not necessarily mean freezing rendering features *tout-court*.

How about rule that all future reserved keywords in stash will start with a hyphen, or some other prefix that users are not allowed to use, much like what happens with `mojo.*`?

**Comment by polettix** [comment]
Let's take this from the other side.

Is there a recommended practice to choose a stash variable name such that there will be no clashes in the future?

Like *start with an uppercase because reserved words will never contain uppercase letters*, or *end with an underscore*, or some other common sense rule that can be adopted without limiting future evolutions?

**Reply by kraih** [reply]
Yes, reserved stash keywords will never contain uppercase letters or start/end with an underscore, or use a `foo.*` prefix.

**Reply by polettix** [reply]
Thanks! The last one seems worse than the other two as `foo.whatever` will not get turned into an easily accessible Perl variable, right?

**Reply by jberger** [reply]
depends on what you mean by worse. I often use dotted stash names in libraries so as to specifically not create a perl variable in templates. you can still get the values in templates using the `stash` helper
